_aicspylibczi/Reader.cpp

```diff
@@ -263,18 +267,39 @@ namespace pylibczi {
 
       imageFactory.setMosaic(isMosaic());
       size_t memOffset = 0;
-      for_each(matches.begin(), matches.end(), [&](const SubblockIndexVec::value_type& match_) {
-          auto subblock = m_czireader->ReadSubBlock(match_.second);
-          const libCZI::SubBlockInfo& info = subblock->GetSubBlockInfo();
-          if (m_pixelType!=info.pixelType)
-              throw PixelTypeException(info.pixelType, "Selected subblocks have inconsistent PixelTypes."
-                                                       " You must select subblocks with consistent PixelTypes.");
 
-          imageFactory.constructImage(subblock->CreateBitmap(),
-              &info.coordinate, info.logicalRect, memOffset, info.mIndex);
+      unsigned int number_of_cores = std::thread::hardware_concurrency();
+      if (number_of_cores - 1 < cores_) {
+          std::cout << "Cores exception requested " << cores_ << " but only " << number_of_cores << " available." << std::endl;
+          throw ThreadingRequestedCoresException(number_of_cores, cores_,
+              "Requested cores should be at most 1 less than the number of available cores.");
+      }
+      {
+          std::vector< std::future<bool> > jobs;
+          Tasks tasks;
+          for_each(matches.begin(), matches.end(), [&](const SubblockIndexVec::value_type match_) {
+              int sb_index = match_.second;
+
+              jobs.push_back(tasks.queue([this, &imageFactory, sb_index, memOffset]()->bool {
+                  auto subblock = m_czireader->ReadSubBlock(sb_index);
+                  const libCZI::SubBlockInfo& info = subblock->GetSubBlockInfo();
+                  if (m_pixelType!=info.pixelType)
+                      throw PixelTypeException(info.pixelType, "Selected subblocks have inconsistent PixelTypes."
+                                                               " You must select subblocks with consistent PixelTypes.");
+
+                  auto bitmap = subblock->CreateBitmap();
+                  libCZI::IntSize size = bitmap->GetSize();
+
+                  imageFactory.constructImage(bitmap, size,
+                      &info.coordinate, info.logicalRect, memOffset, info.mIndex);
+                  return true;
```

Comment:
what happens with the return value?

Reply:
Nothing happens to it in this case but in order to use the task queue the function needs to return something otherwise it isn't considered a task. I wanted to threadpool to handle futures so having a trivial return value enabled me to do it more generally as opposed to only supporting void functions.